error/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+mod internal;
+mod out_point;
+mod script;
+mod spec;
+mod transaction;
+
+pub use block::{BlockError, CellbaseError, CommitError, UnclesError};
+pub use dao::DaoError;
+pub use header::{EpochError, HeaderError, NumberError, PowError, TimestampError};
+pub use internal::InternalError;
+pub use out_point::{CellOutPoint, OutPoint, OutPointError};
+pub use script::ScriptError;
+pub use spec::SpecError;
+pub use transaction::TransactionError;
+
+#[derive(Fail, Debug, Clone, Eq, PartialEq)]
+pub enum ErrorKind {
+    #[fail(display = "OutPointError")]
+    OutPoint,
+    #[fail(display = "TransactionError")]
+    Transaction,
+    #[fail(display = "ScriptError")]
+    Script,
+    #[fail(display = "HeaderError")]
+    Header,
+    #[fail(display = "BlockError")]
+    Block,
+    #[fail(display = "InternalError")]
+    Internal,
+    #[fail(display = "DaoError")]
+    Dao,
+    #[fail(display = "SpecError")]
+    Spec,
+}
+
+#[derive(Debug)]
+pub struct Error {
+    inner: Context<ErrorKind>,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}({})",
+            self.kind(),
+            self.cause().expect("inner cause exist")
+        )
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(kind: ErrorKind) -> Self {
+        Context::new(kind).into()
+    }
+}
+
+impl From<Context<ErrorKind>> for Error {
+    fn from(inner: Context<ErrorKind>) -> Self {
+        Self { inner }
+    }
+}
+
+impl From<OutPointError> for Error {
+    fn from(error: OutPointError) -> Self {
+        error.context(ErrorKind::OutPoint).into()
+    }
+}
+
+impl From<TransactionError> for Error {
+    fn from(error: TransactionError) -> Self {
+        error.context(ErrorKind::Transaction).into()
+    }
+}
+
+impl From<ScriptError> for Error {
+    fn from(error: ScriptError) -> Self {
+        error.context(ErrorKind::Script).into()
+    }
+}
+
+impl From<HeaderError> for Error {
+    fn from(error: HeaderError) -> Self {
+        error.context(ErrorKind::Header).into()
+    }
+}
+
+impl From<BlockError> for Error {
+    fn from(error: BlockError) -> Self {
+        error.context(ErrorKind::Block).into()
+    }
+}
+
+impl From<InternalError> for Error {
+    fn from(error: InternalError) -> Self {
+        error.context(ErrorKind::Internal).into()
+    }
+}
+
+impl From<DaoError> for Error {
+    fn from(error: DaoError) -> Self {
+        error.context(ErrorKind::Dao).into()
+    }
+}
+
+impl From<SpecError> for Error {
+    fn from(error: SpecError) -> Self {
+        error.context(ErrorKind::Spec).into()
+    }
+}
+
+impl From<CapacityError> for Error {
+    fn from(_error: CapacityError) -> Self {
+        InternalError::ArithmeticOverflowCapacity.into()
+    }
+}
+
+impl PartialEq for Error {
+    fn eq(&self, other: &Self) -> bool {
+        if self.kind() != other.kind() {
+            return false;
+        }
+
+        match self.kind() {
+            ErrorKind::OutPoint => {
```

Comment:
what about `self.inner.get_context() == other.inner.get_context()`?

Reply:
`self.inner.get_context()` returns `kind: ErrorKind` only.

When compare `Error`, we need to compare both kinds and the actual errors. So here we compare `self.kind() == other.kind()` , and then downcast and compare the actual errors.

Reply:
Why do we need to compare Error? I think we only need to compare ErrorKind